src/utils/proxySet.ts

```diff
@@ -0,0 +1,89 @@
+import { proxy } from 'valtio'
+
+// properies that we don't want to expose to the end-user
+type InternalProxySet<T> = Set<T> & {
+  data: T[]
+  toJSON: object
+  hasProxy(value: T): boolean
+}
+
+export const proxySet = <T>(initialValues: Iterable<T> | null = []): Set<T> => {
+  const set: InternalProxySet<T> = proxy({
+    data: Array.from(new Set(initialValues)),
+    has(value) {
+      return this.data.indexOf(value) !== -1
+    },
+    hasProxy(value) {
```

Comment:
I wouldn't define this method, but just do it in `add`, which might make the bundle size a little smaller.

And if it's an object we don't need to check `has(value)` which is always false.
Hmmm, let me try:
```js
  add(value) {
    if (typeof value === 'object' && value !== null) {
      value = proxy(value as object)
    }
    if (this.data.indexOf(value) !== -1) {
      this.data.push(value)
    }
    return this
  }
```
Does this work??

Reply:
Oh yes, missed this one. Let me try

Reply:
So i've tried, TS is not super happy. I have to do the following assertion
```ts
if (typeof value === 'object' && value !== null) {
        value = proxy(value as T & object)
      }
      if (this.data.indexOf(value) !== -1) {
        this.data.push(value)
      }
      return this
```

Reply:
that looks fine!

Reply:
My bad xD 

```js
  if (this.data.indexOf(value) === -1) { // instead of !== -1
        this.data.push(value)
      }
```

Reply:
> And if it's an object we don't need to check has(value) which is always false.

This is not 100% true because `Map` and `Set` are objects but they cannot be proxied. So the code for add should be

```tsx
add(value) {
      let hasProxy = false
      if (typeof value === 'object' && value !== null) {
        hasProxy = this.data.indexOf(proxy(value as any)) !== -1
      }
      // this.data.indexOf(value) === -1 or !this.has(value)
      if (this.data.indexOf(value) === -1 && !hasProxy) {
        this.data.push(value)
      }
      return this
    }
```

Reply:
> This is not 100% true because Map and Set are objects

Oh, I missed it. You are absolutely right.

> // add check like instanceof Map && instanceof Set etc...

This doesn't work as `canProxy` **is** customizable. I like the first one in your previous comment.